spec/acceptance/namespaced_factories_spec.rb

```diff
@@ -0,0 +1,87 @@
+describe "Namespaced factories" do
+  before do
+    define_model("Category", name: :string)
+    define_model("Movies::Category", name: :string)
+  end
+
+  context "by default" do
+    before do
+      FactoryBot.define do
+        factory :category do
+          name { "Classics" }
+        end
+
+        with_namespace(:Movies) do
+          factory :category do
+            name { "Emmy Winners" }
+          end
+        end
+      end
+    end
+
+    it "creates a factory prefixed with the namespace" do
+      instance = FactoryBot.build(:movies_category)
+
+      expect(instance).to be_kind_of(Movies::Category)
+      expect(instance.name).to eql("Emmy Winners")
+    end
+
+    it "does not collide with un-namespaced factories" do
+      instance = FactoryBot.build(:category)
+
+      expect(instance).to be_kind_of(Category)
+      expect(instance.name).to eql("Classics")
+    end
+  end
+
+  context "when prefixes are not required" do
+    it "uses the namespaced factory without the prefix" do
+      FactoryBot.define do
+        with_namespace(:Movies, require_prefix: false) do
+          factory :category do
+            name { "Emmy Winners" }
+          end
+        end
+      end
+
+      instance = FactoryBot.build(:category)
+
+      expect(instance).to be_kind_of(Movies::Category)
+      expect(instance.name).to eql("Emmy Winners")
+    end
+
+    it "refuses to clobber existing factories" do
+      FactoryBot.define do
+        factory :category do
+          name { "Classics" }
+        end
+
+        with_namespace(:Movies, require_prefix: false) do
+          factory :category do
+            name { "Emmy Winners" }
+          end
+        end
+      end
+
+      instance = FactoryBot.build(:category)
+
+      expect(instance).to be_kind_of(Category)
```

Comment:
Can you explain why you think this is the correct behavior? As opposed to clobbering the existing factory, or raising a `DuplicateDefinitionError`?

Reply:
Sure, the use case I am currently fiddling with is decomposing a monolith into domain-specific namespaces (and some rails engines); and there's some integration tests that have cross-cutting concerns and also some name collisions.

The integration test suite is currently loading definitions for the following packages:

1. "Core" (the main monolith being decomposed)
2. "ProductA" 
3. "ProductB" 

Each of these have a `User` class; with factory definitions named `:user`; as requiring the developers to use `:product_a_user` within the `ProductA` domain was not particularly desired.

If we clobbered the existing `:user` factory, then existing tests which loaded the `:user` definition from the ProductA module began to fail due to different default values, as well as the reduced surface area of attributes and traits in the ProductA namespace.

Similarly, if we raised the `DuplicateDefinitionError`, our integration tests would stop loading definitions once it attempted to load the `ProductA` `:user` factory.

As a result, discarding the unprefixed duplicate seemed like the safest way to allow folks working within the ProductA/ProductB namespaces the convenience of referencing `create(:user)` while still allowing downstream integration test consumers to  leverage the `:product_a_user` and `:product_b_user` factories.